### ISSUE TITLE
Pattern matching on more than 1 element in a list does not always work

### ISSUE THREAD

**Issue by ChechyLevas** [fable-compiler/Fable#1859]
### Description

When generating a dynamic view, if pattern matching on multiple lists, when events are fired, an error of either "TypeError: _arg1.head is undefined" or "TypeError: _arg1.tail is undefined" is raised.

### Repro code

```F#
module Testing
open Elmish
open Fulma
open Fable.React

type Model = {
    Buttons:Map<int, bool>
}

type Msg =
    |Toggle of int

let init () =
    let buttons = [|1 .. 4|] |> Array.map (fun i -> i, false) |> Map.ofArray
    {Buttons = buttons}, Cmd.none

let update msg model =
    match msg with
    |Toggle i ->
        let newModel =
            match model.Buttons.TryFind i with
            |Some b -> {model with Model.Buttons =  model.Buttons.Add (i, not b)}
            |_ -> model
        newModel, Cmd.none

let makeButton b onClick =
    Button.button
        [ Button.IsFullWidth
          Button.Color IsPrimary
          Button.OnClick onClick ]
        [ str (sprintf "%b" b) ]

let view model dispatch =
    let mapTuple a b = (a, b)
    let rec looper acc = function
        |(i1,b1)::(i2,b2)::rest ->
            let newParent =
                Tile.parent []
                    [
                        Tile.child [] [makeButton b1 (fun _ -> i1 |> Toggle |> dispatch)]
                        Tile.child [] [makeButton b2 (fun _ -> i2 |> Toggle |> dispatch)]
                    ]
            looper (newParent::acc) rest
        |(i,b)::rest ->
            let newParent =
                Tile.parent []
                    [
                        Tile.child [] [makeButton b (fun _ -> i |> Toggle |> dispatch)]
                    ]
            looper (newParent::acc) rest
        |[] -> acc

    let body = model.Buttons |> Map.toList |> looper []

    Tile.ancestor
        [Tile.IsVertical]
        (body)

#if DEBUG
open Elmish.Debug
open Elmish.HMR
#endif

Program.mkProgram init update view2
#if DEBUG
|> Program.withConsoleTrace
#endif
|> Program.withReactBatched "elmish-app"
#if DEBUG
|> Program.withDebugger
#endif
|> Program.run
```
### Expected and actual results

The expected result is to see a 2x2 grid of buttons, each with the text "false", and when a button is clicked, the text is changed to "true".
Instead, when a button is clicked, the above error is raised.
However, if you comment out the first pattern in `looper`, such that only single elements are matched, we get a 4x1 grid of buttons, and the click events are processed correctly with no errors raised. The code to generate the view elements is essentially the same for both cases, which suggests the problem is something to do with how list pattern matching works. Also, the type errors " _arg1.head ..." and " _arg1.tail ...", to my uninitiated nose, smells like some thing to do with lists.
Lastly, you get a "head" error by clicking on a left button, and a "tail" error by clicking on the right button. These correspond to the first and second elements in the pattern respectively, so it seems the position of the element in the pattern determines the precise error raised.

### Related information

* Fable version:
$ yarn info fable-compiler version
yarn info v1.16.0
2.3.12

* Operating system
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.300
 Commit:    73efd5bd87

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.2.300/

Host (useful for support):
  Version: 2.2.5
  Commit:  0a3c9209c0

.NET Core SDKs installed:
  2.2.300 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

**Comment by alfonsogarciacaro** [comment]
Thanks for the report and the code to reproduce the issue @ChechyLevas! After debugging I realized the problem was Fable erases some bindings for optimization (in this case the bindings in the pattern matching for values like `i1` or `i2`) but after optimizing also the tail calls some immutable values become mutable (in this case the list holding the Buttons info). And because here the values were captured in a closure (`fun _ -> i1 |> Toggle |> dispatch`) , when trying to access the list later, this has already become an empty list.

I've disabled the optimization in this case so it should be fixed now. I'll release a new version with the fix shortly. Thanks again for spotting the issue! :+1:

**Comment by ChechyLevas** [comment]
Thanks for the update @alfonsogarciacaro . So why did it work for the single list item pattern? Was that case not subject to the same optimization?

**Comment by alfonsogarciacaro** [comment]
Hmm, not sure what you mean. Do you have an example of a similar pattern that is actually working? The pattern being a value captured by a lambda that also is (or is generated from) a parameter of a tail-call function.

**Comment by ChechyLevas** [comment]
In my repro code above, if you comment out the 2 item pattern from the function `looper`, 
ie: `|(i1,b1)::(i2,b2)::rest -> ...`
so that all you left with is the single item pattern `|(i,b)::rest -> ...`, then the repro code does not produce an error. Is this in line with your expectations?

**Comment by alfonsogarciacaro** [comment]
Not entirely sure why but apparently when you remove the 2 item pattern, the optimization to erase the binding doesn't kick in so the problem doesn't happen :/

I edited a bit your code to turn it into a test. You can [check it in the REPL](https://fable.io/repl/#?code=C4TwDgpgBAogNhAthAdsKBeAUFXUA+sCyaUA9gGZQCGAxsAJZkoBcUAFAK4oPAC0APm68AlFABUUWgAsGcACYAnVG3hJU6OAwDOwHHmSIARhEVRgs7QDoAgvQYA3asAjsx2PJ6iJnM85agAd15pfS8CInU0djpGZgAaKVkFZRQxQTDwvFimFDdMrKgKMjMZOXkoBhQk8tSoeTIahVt7Jxd8rAR0RE5gaiMEGlpaADVqOE5oDCgABk6IdAg4IbjqspTUTEiSYBj7BKalVBEsefQq3g53TIBtfABGKCsrKAAWfABdAgEoG0VFaggKw+MAcCjcSpQQSVRIUcbaCBifA-ACy1DAVkofwBIFOXSgnDA8mc0EQ2gA5hxEGR5Es2GiwAAeKrAeJGMhkOACa6eOijcaTKCMvhDfkTaAAam8FMy+KMW2ptLgVhuZPJH0yDKs1HkFTVHBQZHQRjEiqWeIWUAcDAggW8NKW9R0YF80kwsstyloUDgHMgZj5UAAHu7Cj5gH4Q8ELAU8Ph2Ax7mz7iIWCwEwAmNkZ1MsZS6KECWPhfEoW0ABWoqXQHkKWUdDHJhuUUBuxbrnkd7HB1QA+oXKo9kU7tC6I9IxDcNR2O12e1B+9CGBnviOxzJJ9OZ54tx3fWR-QaK1WNGm+WJ83pCvGGGzc5fC+3PKXj9XQ9u8HOIYufgxV-JnVdTcnzwfdD3YMtAkrasz2GC8IF0dt8CnAc+VOTwzWWYctWAMgABkdHQYcwNMVsvmHBsmxKCALW6agqkqHh0GtW0CSJEl31wfEej6AZSQdZZpguXYTmfS0ANHV1vGlSla3CTChRFQliRcGTvA9RY4FUYgNC2Fi7QU9gJPXN1MNEz9lTsRg2lcE4sB8BjhKtG07WUkksD5MZxVXMBFBZChqgAIgAUgYQKgA&html=DwCwLgtgNgfAUHUBTAhgE3gAm54ElgqbhgAOAtEgI4CuAlgG4C8A5AMID2AdmEj+QBUAnqSQtMAY268erXgA8wAenDQA3JJAoATgGcCTGmABm5ABwssOYLona6pMJl3aJTAEQB9TwAkA8gDKAt5KUHQARrpK2qgSYAB0pNocaDRxdNzxEHRc8QBWuu4wwEq29o5W2DZ2Dk4ubl6+gcGeoRFRMShx5GgcEInJqemZ2bkFRSVltfAlIKgYCMDhKUKSUCi6uh4opBRSPCg5SNpFcDi4aIyYdGgeSFDZuiDkO6Tuaxtb7vePz68TSkuDBmSmWaCE8EQKkgsDgQA&css=BYFwtgNgNAUARgewCYE8AEBvGadoGYIB2IAtAM4CWAXgKYBcaAjAGwAOAHgNwwC+MMAOjABDCoRIBjIiFGEaAJ0zZcSCmVYRhKBngg0uynAHcKSEMAaMADFYCk3XGmA0KAc1CWb9w2gBWAVzIQCjwUSWkaYgYJSJAFB1xhCDdxCjiwMmjY+J9dfRJVeRoJYKJohAh-MEJuPkExVn8QJUdWYSRVQlcGAQAmAFYisAScAmJyanomNgNHEzMLNAGrDhG0EXlXMRJEEBAEMEsh2pggA) and see how the generated JS code is different when you comment out the 2 item pattern.
